pkg/job/discovery.go

```diff
@@ -105,37 +106,58 @@ func runDiscoveryJob(
 		return nil, nil
 	}
 
+	mapResultsToMetricDatas(getMetricDataOutput, getMetricDatas, logger)
+
+	// Remove unprocessed/unknown elements in place, if any. Since getMetricDatas
+	// is a slice of pointers, the compaction can be easily done in-place.
+	getMetricDatas = compact(getMetricDatas, func(m *model.CloudwatchData) bool {
+		return m.MetricID == nil
+	})
+	return resources, getMetricDatas
+}
+
+// mapResultsToMetricDatas walks over all CW GetMetricData results, and map each one with the corresponding model.CloudwatchData.
+//
+// This has been extracted into a separate function to make benchmarking easier.
+func mapResultsToMetricDatas(output [][]cloudwatch.MetricDataResult, datas []*model.CloudwatchData, logger logging.Logger) {
+	// metricIDToData is a support structure used to easily find via a MetricID, the corresponding
+	// model.CloudatchData.
+	metricIDToData := make(map[string]*model.CloudwatchData, len(datas))
+
+	// load the index
+	for _, data := range datas {
+		metricIDToData[*(data.MetricID)] = data
+	}
+
 	// Update getMetricDatas slice with values and timestamps from API response.
 	// We iterate through the response MetricDataResults and match the result ID
 	// with what was sent in the API request.
 	// In the event that the API response contains any ID we don't know about
 	// (shouldn't really happen) we log a warning and move on. On the other hand,
 	// in case the API response does not contain results for all the IDs we've
 	// requested, unprocessed elements will be removed later on.
-	for _, data := range getMetricDataOutput {
+	for _, data := range output {
 		if data == nil {
 			continue
 		}
 		for _, metricDataResult := range data {
-			idx := findGetMetricDataByID(getMetricDatas, metricDataResult.ID)
-			if idx == -1 {
+			// find into index
+			metricData, ok := metricIDToData[metricDataResult.ID]
+			if !ok {
 				logger.Warn("GetMetricData returned unknown metric ID", "metric_id", metricDataResult.ID)
 				continue
 			}
+			// skip elements that have been already mapped but still exist in metricIDToData
+			if metricData.MetricID == nil {
+				continue
+			}
 			// Copy to avoid a loop closure bug
 			dataPoint := metricDataResult.Datapoint
```

Comment:
@cristiangreco this is the pointer reference bug right?

Reply:
Yes!